# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: layout/style/nsComputedDOMStyle.cpp
# Commit: 2ac943ae4424
# Full Hash: 2ac943ae442445f96f5a68d5298b320f30b1313a
# Author: Mats Palmgren <mats@mozilla.com>
# Date: 2019-10-13 09:36:01
# Regressor Bug: 1234311
# File Overlap Count: 1
# Description:
#   Bug 1234311 part 2 - [css-grid-2] Implement 'grid-template-rows/columns' resolved value correctly for subgrid.  r=dholbert
#   
#   This implements the resolution and adds WPTs for:
#   https://github.com/w3c/csswg-drafts/issues/4362
#   
# ==============================================================================

diff -r 43f18d905d75 -r 2ac943ae4424 layout/style/nsComputedDOMStyle.cpp
--- a/layout/style/nsComputedDOMStyle.cpp	Sat Oct 12 17:15:43 2019 +0000
+++ b/layout/style/nsComputedDOMStyle.cpp	Sat Oct 12 17:16:02 2019 +0000
@@ -1548,30 +1548,19 @@
 already_AddRefed<CSSValue> nsComputedDOMStyle::GetGridTemplateColumnsRows(
     const StyleGridTemplateComponent& aTrackList,
     const ComputedGridTrackInfo& aTrackInfo) {
-  if (aTrackList.IsSubgrid()) {
-    // XXX TODO: add support for repeat(auto-fill) for 'subgrid' (bug 1234311)
+  if (aTrackInfo.mIsSubgrid) {
     RefPtr<nsDOMCSSValueList> valueList = GetROCSSValueList(false);
-
-    auto& subgrid = *aTrackList.AsSubgrid();
-
     RefPtr<nsROCSSPrimitiveValue> subgridKeyword = new nsROCSSPrimitiveValue;
     subgridKeyword->SetIdent(eCSSKeyword_subgrid);
     valueList->AppendCSSValue(subgridKeyword.forget());
-
-    auto names = subgrid.names.AsSpan();
-    for (auto i : IntegerRange(names.Length())) {
-      if (MOZ_UNLIKELY(i == subgrid.fill_idx)) {
-        RefPtr<nsROCSSPrimitiveValue> start = new nsROCSSPrimitiveValue;
-        start->SetString(NS_LITERAL_STRING("repeat(auto-fill,"));
-        valueList->AppendCSSValue(start.forget());
-      }
-      AppendGridLineNames(valueList, names[i].AsSpan(),
-                          /*aSuppressEmptyList*/ false);
-      if (MOZ_UNLIKELY(i == subgrid.fill_idx)) {
-        RefPtr<nsROCSSPrimitiveValue> end = new nsROCSSPrimitiveValue;
-        end->SetString(NS_LITERAL_STRING(")"));
-        valueList->AppendCSSValue(end.forget());
-      }
+    for (const auto& lineNames : aTrackInfo.mResolvedLineNames) {
+      AppendGridLineNames(valueList, lineNames, /*aSuppressEmptyList*/ false);
+    }
+    uint32_t line = aTrackInfo.mResolvedLineNames.Length();
+    uint32_t lastLine = aTrackInfo.mNumExplicitTracks + 1;
+    const Span<const StyleCustomIdent> empty;
+    for (; line < lastLine; ++line) {
+      AppendGridLineNames(valueList, empty, /*aSuppressEmptyList*/ false);
     }
     return valueList.forget();
   }